### PR TITLE
Upgrade instructions for coexistence with Docker

### PR DESCRIPTION
**Pull request by starkeepers**
This PR was requested by @tomponline  on the [linuxcontainers forum](https://discuss.linuxcontainers.org/t/lxd-and-docker-firewall-redux-how-to-deal-with-forward-policy-set-to-drop/9953/12).

### CODE REVIEW

File: doc/howto/network_bridge_firewalld.md [hunk 1]
```diff
@@ -98,6 +98,12 @@ A common reason for these issues is that Docker sets the FORWARD policy to `drop
 See [Docker on a router](https://docs.docker.com/network/iptables/#docker-on-a-router) for detailed information.
 
 The easiest way to prevent such issues is to uninstall Docker from the system that runs LXD.
-If that is not an option, use the following command to explicitly allow network traffic from your network bridge to your external network interface:
+If that is not an option, use the following command to explicitly allow network traffic between your network bridge and your external network interface:
 
     iptables -I DOCKER-USER -i <network_bridge> -o <external_interface> -j ACCEPT
+    iptables -I DOCKER-USER -i <external_interface> -o <network_bridge> -j ACCEPT
+
+To make it stick across reboots, the above can be added to `/etc/network/interfaces` along the lines of:
```
Comment (ru-fu) [inline doc/howto/network_bridge_firewalld.md:106]:
```suggestion
To make the configuration persist across reboots, add the commands to `/etc/network/interfaces`:
```

**Review by ru-fu — COMMENTED** [review]
Thanks!

File: doc/howto/network_bridge_firewalld.md [hunk 2]
```diff
@@ -98,6 +98,12 @@ A common reason for these issues is that Docker sets the FORWARD policy to `drop
 See [Docker on a router](https://docs.docker.com/network/iptables/#docker-on-a-router) for detailed information.
 
 The easiest way to prevent such issues is to uninstall Docker from the system that runs LXD.
-If that is not an option, use the following command to explicitly allow network traffic from your network bridge to your external network interface:
+If that is not an option, use the following command to explicitly allow network traffic between your network bridge and your external network interface:
 
     iptables -I DOCKER-USER -i <network_bridge> -o <external_interface> -j ACCEPT
+    iptables -I DOCKER-USER -i <external_interface> -o <network_bridge> -j ACCEPT
```
Comment (tomponline) [inline doc/howto/network_bridge_firewalld.md:104]: I'm concerned by this being needed as it allows all inbound traffic from the external interface onto the lxdbr0 bridge.
Why is this needed?

Our aim for the instructions on this page is to allow containers connected to lxdbr0 to get automatic IP configuration from LXD's DHCP server, DNS fron LXD's DNS server and outbound internet access.

Comment (tomponline) [reply doc/howto/network_bridge_firewalld.md:104]: I'd like to understand what is being blocked that is not covered by the existing rules, and then maybe we can suggest a modification that is less drastic.

**Review by tomponline — CHANGES_REQUESTED** [review]
See inline comments.

File: doc/howto/network_bridge_firewalld.md (hunk 2 above)
Comment (starkeepers) [reply doc/howto/network_bridge_firewalld.md:104]: > what is being blocked that is not covered by the existing rules

Eg. from [the linuxcontainers thread above](https://discuss.linuxcontainers.org/t/lxd-and-docker-firewall-redux-how-to-deal-with-forward-policy-set-to-drop/9953/3) "internet access from lxc containers stop working". Similar problems are reported in [various](https://discuss.linuxcontainers.org/t/containers-fails-to-reach-internet/14671) [other](https://discuss.linuxcontainers.org/t/cant-get-ipv4-addresses-with-docker-installed/8019/8) [places](https://discuss.linuxcontainers.org/t/lxd-4-12-container-networking-failure-with-docker/10487/13) as well is in [a docker issue](https://github.com/docker/for-linux/issues/103). In several of those places the documentation page in this PR is offered as a solution - but it doesn't actually cure the ping/nginx/postfix/etc problem, whereas making it bidirectional does cure it (at least on my freshly installed debian bookworm).

In the absence of solution, users resort to an even more radical solution in the above threads, eg. `iptables -F FORWARD && iptables -P FORWARD ACCEPT`. Offering a more restrictive solution helps that, and if you can offer an even more restricted solution that still allows i/o networking from the containers in the presence of docker, that would be great.

Comment (tomponline) [reply doc/howto/network_bridge_firewalld.md:104]: Are you saying that egress traffic doesn't work without these rules either, or that in your specific scenario you need ingress traffic to work as well?

If egress traffic doesn't work then I think we need to explore specifically which traffic is being blocked, as it shouldn't need to allow all ingress traffic from the external traffic in order to get egress traffic working (we should only need to allow traffic related to egress connections).

If you need ingress traffic to work then I would suggest we add this as a separate section/block so that people who only need egress traffic to work aren't accidentally allow ingress traffic and reducing their network security.

Comment (tomponline) [reply doc/howto/network_bridge_firewalld.md:104]:
```
iptables -I DOCKER-USER -i <external_interface> -o <network_bridge> -j ACCEPT
```

Will potentially expose the LXD bridge to the external network, especially as LXD enables routing by default.
I think we should only recommend firewall commands that enable traffic from containers to the LXD bridge and egress to the external network.

Comment (tomponline) [reply doc/howto/network_bridge_firewalld.md:104]: You mention the "but it doesn't actually cure the ping/nginx/postfix/etc problem" can you expand on what precisely that is?

Are you listing the services running inside the containers that you're trying to expose to the external network?

**Comment by stgraber** [comment]
Marking as Incomplete as there are open questions from @tomponline

**Comment by starkeepers** [comment]
> Marking as Incomplete as there are open questions from @tomponline

@stgraber  I'm not sure what the change is that is being requested?

**Comment by tomponline** [comment]
Basically the line `iptables -I DOCKER-USER -i <external_interface> -o <network_bridge> -j ACCEPT` is not something I would want to add to our docs as recommended for getting LXD instances to work when Docker is installed because it opens up the managed LXD bridge to the external interface. 

It also shouldn't be needed to allow LXD instances to use outbound network connectivity and get DHCP/DNS from the managed bridge itself.

My questions are above:

- You mention the "but it doesn't actually cure the ping/nginx/postfix/etc problem" can you expand on what precisely that is?
- Are you listing the services running inside the containers that you're trying to expose to the external network?

I am trying to figure out whether the change you're requesting is to allow egress traffic and DHCP/DNS or whether it is to allow services running inside LXD instances to be reachable from the external interface.

**Comment by stgraber** [comment]
Closing for inactivity
